### DISCUSSION TITLE
PIO SM_ADDR has unexpected value

### DISCUSSION THREAD

**Discussion by kc64** [micropython/micropython#15102] [RP2040 / Pico]
I have a test program consisting of only the instructions to attempt to understand the behavior I'm seeing:

```
wait(1, gpio, 15)
wait(0, gpio, 15)
```

I also have the GPIO set as an input with pullup.

I have a pushbutton to toggle the GPIO and then inspect with:

```
PIO0_BASE = const(0x50200000)
SM0_ADDR = const(0x0d4)
SM0_INSTR = const(0x0d8)

def a():
    print(hex(mem32[PIO0_BASE + SM0_ADDR]), hex(mem32[PIO0_BASE + SM0_INSTR]))

>>> a()
0x1e 0x208f
>>> a()
0x1f 0x200f
```

Why are these instructions at the bottom of the instruction memory? I fully expected them to be at 0x00 and 0x01.

**Comment by GitHubsSilverBullet** [comment]
Because the ›InstructionMemory‹ gets filled backwards.

**Reply by kc64** [reply]
Interesting! I was examining the source code and couldn't find any evidence that was happening, but my C++ skills are not impressive. Many thanks for your insight and quick reply.
